lib/puppet/transaction/report.rb

```diff
@@ -158,6 +168,21 @@ def compute_status(resource_metrics, change_metric)
   end
 
   # @api private
+  def has_noop_change_events?(resource)
+    resource.events.any? { |event| (event.status == 'noop' && event.name == :content_changed) }
```

Comment:
Afaik, all events are "changes" or "proposed changes" @kbrezina, (the exception being audit events). The presence of an event suggests there was a change. You shouldn't need to compare previous/desired.